### PR TITLE
Df.at impl

### PR DESCRIPTION
**Pull request by 1e-to**

### CODE REVIEW

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 1]
```diff
@@ -1759,13 +1759,76 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+                       '    if self._dataframe.index[i] == row:',
+                       '      check_row = True',
+                       '  if check_row:',
+                       f'    data = self._dataframe._data[{col_idx}]',
+                       '    res_data = pandas.Series(data, index=self._dataframe.index)',
+                       '    return res_data.at[row]',
+                       "  raise IndexingError('Index is out of bounds for axis')"]
+
+    else:
+        raise IndexingError('Index is out of bounds for axis')
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: I propose to change the code a bit
```suggestion
    for i in range(len(self.columns)):
        if self.columns[i] == col:
            func_lines += [
                '  row, _ = idx',
                '  check_row = False',
                '  for i in prange(len(self._dataframe.index)):',
                '    if self._dataframe.index[i] == row:',
                '      check_row = True',
                '  if check_row:',
                f'    data = self._dataframe._data[{i}]',
                '    res_data = pandas.Series(data, index=self._dataframe.index)',
                '    return res_data.at[row]',
                '  raise IndexingError("Index is out of bounds for axis")'
            ]
            break
    else:
        raise IndexingError('Index is out of bounds for axis')
```

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 2]
```diff
@@ -1759,13 +1759,76 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
 def sdc_pandas_dataframe_accessor_getitem(self, idx):
     if not isinstance(self, DataFrameGetitemAccessorType):
         return None
 
     accessor = self.accessor.literal_value
 
+    if accessor == 'at':
+        if isinstance(idx, types.Tuple) and isinstance(idx[1], types.Literal):
+            row = idx[0]
+            col = idx[1].literal_value
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py]: Shouldn't we add limitation about `column` supported only to be literal?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 3]
```diff
@@ -1759,13 +1759,73 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
 @sdc_overload(operator.getitem)
 def sdc_pandas_dataframe_accessor_getitem(self, idx):
     if not isinstance(self, DataFrameGetitemAccessorType):
         return None
 
     accessor = self.accessor.literal_value
 
+    if accessor == 'at':
+        if isinstance(idx, types.Tuple) and isinstance(idx[1], types.Literal):
+            row = idx[0]
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py:1823]: Need to check that first item in idx is of the same type as index and second item is string

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 4]
```diff
@@ -1759,13 +1759,80 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
+            row, _ = idx
+            check_row = False
+            for i in prange(len(self._dataframe.index)):
+                if self._dataframe.index[i] == row:
+                    check_row = True
+            if check_row:
+                data = self._dataframe._data[2]
+                res_data = pandas.Series(data, index=self._dataframe.index)
+                return res_data.at[row]
+            raise IndexingError('Index is out of bounds for axis')
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py:1775]: For which axis?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 5]
```diff
@@ -1759,13 +1759,80 @@ def _df_getitem_unicode_idx_impl(self, idx):
     ty_checker.raise_exc(idx, expected_types, 'idx')
 
 
+def df_getitem_tuple_at_codegen(self, row, col):
+    """
+    Example of generated implementation:
+        def _df_getitem_tuple_at_impl(self, idx):
+            row, _ = idx
+            check_row = False
+            for i in prange(len(self._dataframe.index)):
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py:1768]: Do we actually need this? Wouldn't it be checked in Series.at?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 6]
```diff
@@ -245,9 +245,14 @@ def hpat_pandas_series_loc_impl(self, idx):
...
             def hpat_pandas_series_at_impl(self, idx):
                 index = self._series.index
+                check = False
                 mask = numpy.empty(len(self._series._data), numpy.bool_)
                 for i in numba.prange(len(index)):
                     mask[i] = index[i] == idx
+                    if mask[i] == True:  # noqa
+                        check = True
+                if check != True:  # noqa
+                    raise ValueError("Index is not in the DataFrame")
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py:255]: Error message doesn't feel right. It is Series method, not DataFrame

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 7]
```diff
@@ -1876,6 +1876,37 @@ def _df_getitem_unicode_idx_impl(self, idx):
...
 
 
+def df_getitem_tuple_at_codegen(self, row, col):
+    """
+    Example of generated implementation:
+        def _df_getitem_tuple_at_impl(self, idx):
+            row, _ = idx
+            data = self._dataframe._data[2]
+            res_data = pandas.Series(data, index=self._dataframe.index)
+            return res_data.at[row][0]
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py:1886]: And what if Series returned more than one result?

File: sdc/datatypes/hpat_pandas_dataframe_functions.py [hunk 8]
```diff
@@ -2181,6 +2234,57 @@ def sdc_pandas_dataframe_iat_impl(self):
...
 
+@sdc_overload_attribute(DataFrameType, 'at')
+def sdc_pandas_dataframe_at(self):
+    """
+    Intel Scalable Dataframe Compiler User Guide
+    ********************************************
+
+    Limitations
+    -----------
+    - Parameter ``column`` in ``idx`` must be a literal value.
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_dataframe_functions.py:2245]: Please add limitation, that `at` always returns Series

**Comment by AlexanderKalistratov** [comment]
@1e-to conflicts

**Pull request merged**